client/go/internal/vespa/document/dispatcher.go

```diff
@@ -5,59 +5,125 @@ import (
 	"sync"
 )
 
+const maxAttempts = 10
+
 // Dispatcher dispatches documents from a queue to a Feeder.
 type Dispatcher struct {
-	concurrencyLevel int
-	feeder           Feeder
-	pending          chan Document
-	closed           bool
-	mu               sync.RWMutex
-	wg               sync.WaitGroup
+	workers  int
+	feeder   Feeder
+	ready    chan Id
+	inflight map[string]*documentGroup
+	mu       sync.RWMutex
+	wg       sync.WaitGroup
+	closed   bool
+}
+
+// documentGroup holds document operations which share their ID, and must be dispatched in order.
+type documentGroup struct {
+	id         Id
+	failed     bool
+	operations []documentOp
+	mu         sync.Mutex
+}
+
+type documentOp struct {
+	document Document
+	attempts int
+}
+
+func (g *documentGroup) append(op documentOp) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	g.operations = append(g.operations, op)
 }
 
-func NewDispatcher(feeder Feeder, concurrencyLevel int) *Dispatcher {
-	if concurrencyLevel < 1 {
-		concurrencyLevel = 1
+func NewDispatcher(feeder Feeder, workers int) *Dispatcher {
+	if workers < 1 {
+		workers = 1
 	}
 	d := &Dispatcher{
-		concurrencyLevel: concurrencyLevel,
-		feeder:           feeder,
-		pending:          make(chan Document, 4*concurrencyLevel),
+		workers:  workers,
+		feeder:   feeder,
+		inflight: make(map[string]*documentGroup),
 	}
-	d.readPending()
+	d.start()
 	return d
 }
 
-func (d *Dispatcher) readPending() {
-	for i := 0; i < d.concurrencyLevel; i++ {
+func (d *Dispatcher) dispatchAll(g *documentGroup) int {
+	g.mu.Lock()
+	defer g.mu.Unlock()
```

Comment:
Hmm, won't this block the client thread if it tries to add another document with this ID, until all operations for that ID have completed? I guess it's not likely to be a problem ... but is the locking necessary?